staging/src/k8s.io/apiserver/pkg/storage/cacher/store_btree.go

```diff
@@ -385,3 +405,84 @@ func (i *indexer) delete(key, value string, index map[string]map[string]*storeEl
 		delete(index, value)
 	}
 }
+
+// newStoreSnapshotter returns a storeSnapshotter that stores snapshots for serving read requests with exact resource versions (RV) and pagination.
+//
+// Snapshots are created by calling Clone method on orderedLister, which is expected to be fast and efficient thanks to usage of B-trees.
+// B-trees can create a lazy copy of the tree structure, minimizing overhead.
+//
+// Assuming the watch cache observes all events and snapshots cache after each of them,
+// requests for a specific resource version can be served by retrieving the snapshot with the greatest RV less than or equal to the requested RV.
+// To make snapshot retrivial efficient we need an ordered data structure, such as tree.
+//
+// The initial implementation uses a B-tree to achieve the following performance characteristics:
+//   - `Add`: Adds a new snapshot. Complexity: O(log n). Executed for each watch event observed by the cache.
+//   - `GetLessOrEqual`: Retrieves the snapshot with the greatest RV less than or equal to the requested RV. Complexity: O(log n). Executed for each LIST request with match=Exact or continuation.
+//   - `RemoveLessOrEqual`: Cleans up snapshots outside the watch history window. Complexity: O(n log n). Executed per watch event observed when the cache is full.
+//   - `Reset`: Cleans up all snapshots. Complexity: O(log n). Executed when the watch cache is reinitialized.
```

Comment:
Why it's O(log n)?

Reply:
It came from option in btree to reuse nodes. Btree Reset method allows caller to reuse nodes for free node list and avoid reallocation. Disabled it for now to prevent complications.